src/sources/util/decoding/config.rs

```diff
@@ -0,0 +1,146 @@
+//! Definition of a common decoding configuration which is able to resolve
+//! `Decoder` implementations that are registered in the global inventory.
+
+#![deny(missing_docs)]
+
+use super::{BytesDecoder, Decoder};
+use crate::event::Value;
+use bytes::Bytes;
+use lazy_static::lazy_static;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+lazy_static! {
+    /// A hashmap that resolves from shorthand name to `Decoder` implementation.
+    ///
+    /// The `Decoder` implementations must register themselves in the global
+    /// inventory using `inventory::submit!` to be resolved at this point.
+    static ref DECODERS: HashMap<&'static str, &'static dyn super::Decoder> =
+        inventory::iter::<Box<dyn super::Decoder>>
+            .into_iter()
+            .map(|decoder| (decoder.name(), &**decoder))
+            .collect();
+}
+
+/// A decoding configuration, either by shorthand name or fully resolved with
+/// options, e.g:
+///
+/// ```toml
+/// # Decoder by shorthand name.
+/// decoding = "json"
+/// ```
+///
+/// ```toml
+/// # Decoder with options.
+/// decoding = { codec = "json", drop_invalid = true }
+/// ```
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum DecodingConfig {
+    /// A shorthand name which is used to resolve the decoder from the
+    /// global inventory.
+    Name(String),
+    /// A fully resolved decoder with options applied.
+    Decoder(Box<dyn super::Decoder>),
```

Comment:
I don't see this variant being constructed anywhere in this PR. Did I miss something?

Reply:
This happens somewhat magically, by the combined usage of `#[serde(untagged)]` and `#[typetag::serde(tag = "codec")]`.

The `config_codec_with_options ` unit test constructs this variant.